### PR TITLE
Add warnings/errors for unreasonable DHW temperatures for WaterUse:Equipment

### LINKED ISSUE #7133: Feature Request: Generate Warning when a WaterUse:Equipment object has a target temperature is below water temperature
Issue overview
--------------
In the FullServiceRestaurant model created by the Create DOE Prototype Building BCL measure, one of the WaterUse:Equipment objects has a target temperature of 15.5F = -9.1C, a.k.a. ice. 

In this case, there isn't a problem with having not frozen water coming out of a faucet. However, in other instances it would be good to have a warning generated if the faucet temperature is above the target temperature of the WaterUse:Equiptment object. At the very least, this would be good to catch human input error.

### Details
Some additional details for this issue (if relevant):
 - Wnidows 10
 - EnergyPlus 9.0

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [x] Defect file added https://drive.google.com/file/d/1rPdD9HkG_tevBclcMMXTfDTbMoWVYWXq/view?usp=sharing
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### LINKED ISSUE #8806: No error/warning if DHW hot water temp is less than the cold water temp
Issue overview
--------------
If a user accidentally sets the hot water temperature to a value below the cold water temperature in the WaterUse:Equipment object, WaterUse:Connections object, or PlantLoop outlet temperature, there is no error or warning and if the Target Temperature is above the Hot Water Temperature, the Hot Water Mass Flow Rate will be set to it's peak flow rate.  This results in unnecessary high flow rates and no Cold Water Mass Flow Rate.  If the Target Temperature is below the Hot Water Temperature, you will end up with a negative Hot Water Mass Flow Rate.  This was mentioned in the code below:

```
        // Calculate hot and cold water mixing at the tap
        if (this->TotalMassFlowRate > 0.0) {
            // Calculate the flow rates needed to meet the target temperature
            if (this->HotTemp == this->ColdTemp) { // Avoid divide by zero
                // There is no hot water
                this->HotMassFlowRate = 0.0;

                // Need a special case for HotTemp < ColdTemp, due to bad user input  (but could happen in a plant loop accidentally)

            } else if (this->TargetTemp > this->HotTemp) {
                this->HotMassFlowRate = this->TotalMassFlowRate;

            } else {
                this->HotMassFlowRate = this->TotalMassFlowRate * (this->TargetTemp - this->ColdTemp) / (this->HotTemp - this->ColdTemp);
            }
```

### Details
Some additional details for this issue (if relevant):
 - EnergyPlus v9.5

### Checklist
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Defect file added (list location of defect file here)
 - [ ] Ticket added to Pivotal for defect (development team task)
 - [ ] Pull request created (the pull request will have additional tasks related to reviewing changes that fix this defect)

### PR DESCRIPTION
**Pull request by matthew-larson**
Pull request overview
---------------------
 - Fixes #7133 and #8806 
 - This pull request adds warnings for unreasonable DHW temperatures for the target temperature, hot water temperature, and cold water temperature.  A general warning will be included if it only occurs once, but a recurring warning will be generated if it occurs multiple times.  This will help prevent unreasonable water temperatures by warning the user if the hot water temperature is below the cold water temperature, and if the target temperature is below the cold water temperature or above the hot water temperature, making the target temperature unachievable and the energy consumption inaccurate.
 - Created a new unit test file for WaterUse.unit.cc that has a unit test that checks these three warnings.

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally

### CODE REVIEW

**Comment by matthew-larson** [comment]
All diffs are due to situations where the target water temperature is unachievable due to the hot water temperature, or the hot water temperature is below the cold water temperature.  Not sure if these should be investigated in the models to eliminate these warnings.

File: src/EnergyPlus/WaterUse.cc [hunk 1]
```diff
@@ -1005,18 +1007,73 @@ namespace WaterUse {
...
+                                      "...Hot water temperature should be greater than or equal to the cold water temperature. "
+                                      "Verify temperature setpoints and schedules.");
+                } else {
+                    ShowRecurringWarningErrorAtEnd(
+                        state,
+                        this->Name + "\" - Hot water temperature should be greater than or equal to the cold water temperature error continues...",
+                        this->HWTempErrIndex,
+                        this->HotTemp,
+                        this->HotTemp);
+                }
```
Comment (Nigusse) [inline src/EnergyPlus/WaterUse.cc:1140]: Catches when *Cold* water temp is > available *Hot* water temp and throws one time and recurring warning error message.

File: src/EnergyPlus/WaterUse.cc [hunk 2]
```diff
@@ -1005,18 +1007,73 @@ namespace WaterUse {
...
+                                      "...Target water temperature should be less than or equal to the hot water temperature. "
+                                      "Verify temperature setpoints and schedules.");
+                } else {
+                    ShowRecurringWarningErrorAtEnd(
+                        state,
+                        this->Name + "\" - Target water temperature should be less than or equal to the hot water temperature error continues...",
+                        this->TargetTempErrIndex,
+                        this->HotTemp,
+                        this->ColdTemp);
+                }
```
Comment (Nigusse) [inline src/EnergyPlus/WaterUse.cc:1162]: Catches when *Target* water temp is > available *Hot* water temp and throws one time and recurring warning error message.

File: src/EnergyPlus/WaterUse.cc [hunk 3]
```diff
@@ -1005,18 +1007,73 @@ namespace WaterUse {
...
+                                      "...Target water temperature should be greater than or equal to the cold water temperature. "
+                                      "Verify temperature setpoints and schedules.");
+                } else {
+                    ShowRecurringWarningErrorAtEnd(
+                        state,
+                        this->Name + "\" - Target water temperature should be greater than or equal to the cold water temperature error continues...",
+                        this->TargetTempErrIndex,
+                        this->HotTemp,
+                        this->ColdTemp);
+                }
```
Comment (Nigusse) [inline src/EnergyPlus/WaterUse.cc:1188]: Catches when *Target* water temp is < available *Cold* water temp and throws one time and recurring warning error message.

File: tst/EnergyPlus/unit/WaterUse.unit.cc [hunk 4]
```diff
@@ -0,0 +1,494 @@
...
+        "   ** Warning ** CalcEquipmentFlowRates: Target water temperature is less than the cold water temperature",
+        "   **   ~~~   ** ...target water temperature       = 0.000 C",
+        "   **   ~~~   ** ...cold water temperature       = 15.000 C",
+        "   **   ~~~   ** ...Note: target water temperature should be greater than or equal to the cold water temperature",
+        "   **   ~~~   ** ...Target water temperature should be greater than or equal to the cold water temperature. Verify temperature setpoints "
+        "and schedules.",
+    });
+
+    EXPECT_TRUE(compare_err_stream(error_string3, true));
+}
```
Comment (Nigusse) [inline tst/EnergyPlus/unit/WaterUse.unit.cc:498]: This unit test demonstrates the three conditions where throwing warning error message is warranted. Update the branch, built and ran the unit test locally. Also ran the unit test suite locally and all passed.

**Review by Nigusse — APPROVED** [review]

**Comment by Nigusse** [comment]
Now there are 35 example files that throw warning error message because they are caught by the fix. But these are expected error messages unless we want to modify the example files to eliminate the warning error. @Myoldmopar

**Comment by Nigusse** [comment]
If the warning error message is a concern another alternative is to not report the warning error message unless "DisplayExtraWarnings" variable is set to true.

**Comment by mjwitte** [comment]
@matthew-larson @Nigusse Ultimately, we want to fix these so the example files deliver the requested target temp without warnings. I would not hide these under "DisplayExtraWarnings". That could happen here or as a sperate defect/PR.

These are are all good warnings that indicate the target temp won't be met, and why. Do we have a warning that checks the supply temp vs target temp directly and warns that it didn't deliver? Or would that even be useful? If all of these checks pass, are there other things that could prevent the target temp from being met?

**Comment by matthew-larson** [comment]
@mjwitte @Nigusse I ended up not doing a check for supply temp vs target temp as this would involve whole different part of the code and I thought would probably be better as a separate issue/pull request.  This could probably go along with DHW heating capacity checks which could be the other reason that the target temp wouldn't be reached.

I can go ahead and create another issue/pull request to fix the example files so we can push this one through sooner than later without me holding it up.

**Comment by Nigusse** [comment]
@mjwitte @matthew-larson  I suspect sometimes the available capacity and demand mismatch may lead to hot water temperature to drop below the target temperature. But these needs to investigate case by case. I think removing the error message from the example files should be handled in a separate issue.

**Comment by Nigusse** [comment]
@mjwitte I was making comment but ended up closing the issue. That was inadvertent.

**Comment by Nigusse** [comment]
I looked at a couple of example files trying to understand the source and causes of the warning messages. For instance, the example file **ReportHeatEmission_RefFSRestaurant.idf** has a total of 41 warning massages and these messages occur during warmup period only. There are also two other observations I made:
- the warning was issued at Simulation time=07/21 00:00 - 00:15
- when I switched the order of the SizingPeriod:DesignDay (summer first) the warning error message occurred at Simulation time=01/21 00:00 - 00:15.  
- examining the hot water temperature in the tank during these hours indicate the hot water temperature was ~59C, way higher than the target temperature 49C. 
- This implies there is initialization problem.

@matthew-larson please add timestamp reporting function ` ShowContinueErrorTimeStamp(state, "") ` where you added the warning error message code (three locations). This will help us determine if the problem I am seeing is the same across all the example files with error file diffs and it is helpful to have timestamp for future use.

File: src/EnergyPlus/WaterUse.cc [hunk 5]
```diff
@@ -1005,18 +1007,73 @@ namespace WaterUse {
...
-                // Need a special case for HotTemp < ColdTemp, due to bad user input  (but could happen in a plant loop accidentally)
-
-            } else if (this->TargetTemp > this->HotTemp) {
+            } else if ((this->ColdTemp - this->HotTemp) > EPSILON) {
+                // Special case for HotTemp < ColdTemp, due to bad user input (could happen in a plant loop accidentally)
+                this->HotMassFlowRate = 0;
+                // print error for variables of hot water temperature
+                ++this->HWTempErrorCount;
+                if (this->HWTempErrorCount < 2) {
+                    ShowWarningError(state, "CalcEquipmentFlowRates: Hot water temperature is less than the cold water temperature");
```
Comment (Nigusse) [inline src/EnergyPlus/WaterUse.cc:1125]: Add this `ShowContinueErrorTimeStamp(state, "");` here

File: src/EnergyPlus/WaterUse.cc [hunk 6]
```diff
@@ -1005,18 +1007,73 @@ namespace WaterUse {
...
+                        this->HotTemp);
+                }
+            } else if ((this->TargetTemp - this->HotTemp) > EPSILON) {
+                // Target temp is hotter than the hot water temp; can't meet target temperature
                 this->HotMassFlowRate = this->TotalMassFlowRate;
-
+                // print error for variables of target water temperature
+                ++this->TargetTempErrorCount;
+                if (this->TargetTempErrorCount < 2) {
+                    ShowWarningError(state, "CalcEquipmentFlowRates: Target water temperature is greater than the hot water temperature");
```
Comment (Nigusse) [inline src/EnergyPlus/WaterUse.cc:1147]: Add this `ShowContinueErrorTimeStamp(state, "");` here

File: src/EnergyPlus/WaterUse.cc [hunk 7]
```diff
@@ -1005,18 +1007,73 @@ namespace WaterUse {
...
                 this->HotMassFlowRate = this->TotalMassFlowRate * (this->TargetTemp - this->ColdTemp) / (this->HotTemp - this->ColdTemp);
             }
 
             if (this->HotMassFlowRate < 0.0) {
                 // Target temp is colder than the cold water temp; don't allow colder
                 this->HotMassFlowRate = 0.0;
+                // print error for variables of target water temperature
+                ++this->TargetTempErrorCount;
+                if (this->TargetTempErrorCount < 2) {
+                    ShowWarningError(state, "CalcEquipmentFlowRates: Target water temperature is less than the cold water temperature");
```
Comment (Nigusse) [inline src/EnergyPlus/WaterUse.cc:1173]: Add this `ShowContinueErrorTimeStamp(state, "");` here

**Comment by matthew-larson** [comment]
Thanks @Nigusse for your review.  I went ahead and added the `ShowContinueErrorTimeStamp(state, "");` lines as suggested, updated the unit tests accordingly, and merged in the latest develop branch.

**Comment by Myoldmopar** [comment]
I'll pull and test this locally.

**Comment by Myoldmopar** [comment]
Builds and passes locally with develop pulled in.  Regressions locally are exactly what it was showing the last time Decent ran.  Merging, thanks @matthew-larson and @Nigusse!

**Comment by matthew-larson** [comment]
Thanks @Myoldmopar!

**Pull request merged**